AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/RootContextHandler.java

```diff
@@ -26,55 +24,63 @@
 import org.androidannotations.ElementValidation;
 import org.androidannotations.annotations.RootContext;
 import org.androidannotations.handler.BaseAnnotationHandler;
+import org.androidannotations.handler.MethodInjectionHandler;
 import org.androidannotations.helper.CanonicalNameConstants;
+import org.androidannotations.helper.InjectHelper;
 import org.androidannotations.holder.EBeanHolder;
 
 import com.helger.jcodemodel.AbstractJClass;
 import com.helger.jcodemodel.IJExpression;
 import com.helger.jcodemodel.JBlock;
-import com.helger.jcodemodel.JConditional;
-import com.helger.jcodemodel.JInvocation;
+import com.helger.jcodemodel.JExpr;
+import com.helger.jcodemodel.JOp;
 
-public class RootContextHandler extends BaseAnnotationHandler<EBeanHolder> {
+public class RootContextHandler extends BaseAnnotationHandler<EBeanHolder>implements MethodInjectionHandler<EBeanHolder> {
+
+	private final InjectHelper<EBeanHolder> injectHelper;
 
 	public RootContextHandler(AndroidAnnotationsEnvironment environment) {
 		super(RootContext.class, environment);
+		injectHelper = new InjectHelper<>(validatorHelper, this);
 	}
 
 	@Override
 	public void validate(Element element, ElementValidation validation) {
-		validatorHelper.enclosingElementHasEBeanAnnotation(element, validation);
+		injectHelper.validate(RootContext.class, element, validation);
 
-		validatorHelper.extendsContext(element, validation);
+		Element param = injectHelper.getParam(element);
+		validatorHelper.extendsContext(param, validation);
 
 		validatorHelper.isNotPrivate(element, validation);
 	}
 
 	@Override
 	public void process(Element element, EBeanHolder holder) {
-		String fieldName = element.getSimpleName().toString();
+		injectHelper.process(element, holder);
+	}
+
+	@Override
+	public JBlock getInvocationBlock(EBeanHolder holder) {
+		return holder.getInitBodyInjectionBlock();
+	}
 
-		TypeMirror elementType = element.asType();
+	@Override
+	public IJExpression getInstanceInvocation(Element element, EBeanHolder holder, Element param) {
+		TypeMirror elementType = param.asType();
 		String typeQualifiedName = elementType.toString();
 
-		JBlock body = holder.getInitBodyInjectionBlock();
 		IJExpression contextRef = holder.getContextRef();
 
 		if (CanonicalNameConstants.CONTEXT.equals(typeQualifiedName)) {
-			body.assign(ref(fieldName), contextRef);
+			return contextRef;
 		} else {
 			AbstractJClass extendingContextClass = getEnvironment().getJClass(typeQualifiedName);
-			JConditional cond = body._if(holder.getContextRef()._instanceof(extendingContextClass));
```

Comment:
i think that would generate wired code for methods with multiple parameters. as the elements are not processed in a logical order - similar to the calling order for `@AfterXxx`methods?

if i'm not wrong, this could result in something like this:
```java
if(!(bla instanceof SomeContext)) {
   // logging
}
callInjecteion(xyz); // this is another injection.
if(!(bla2 instanceof AnotherContext)) {
   // logging
}
callContextInjection(bla, bla2);
```

while it wont make sense to inject more than one context, it is possible.